app/services/edition_assertions.rb

```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module EditionAssertions
+  class Error < StandardError
+    attr_reader :edition
+
+    def initialize(edition, assertion)
+      @edition = edition
+      super("Assertion failed: #{assertion}")
+    end
+  end
+
+  class StateError < Error; end
+
+  def assert_edition_state(edition, options = {}, &block)
+    return if block.call(edition)
+
+    assertion = options[:assertion] || block.inspect
```

Comment:
I wonder if we actually have a need for custom messaging? I'd also be keen if didn't use `block.inspect` since it's either hard or impossible to distinguish.

I think we'd be ok with just a generic message of "Edition does not meet state requirements"

Reply:
The current mechanism has a small benefit, but perhaps it's unclear:

```
assert_edition_state(edition, &:editable?)
=> "Assertion failed: Proc<dfdfdfsdf>(:editable?)"

assert_edition_state(edition) { |e| e.editable? }
=> "Assertion failed: Proc<dfsdfdf>"
```

So the `assertion:` option is to support messaging for more custom blocks. Especially where we have complex assertions, I think it's helpful to write what we're actually asserting.

In order to make the logs useful, I think we should have something non-static in them. The stacktrace seems overkill for a non-error. I guess this approach has the problem of it being unclear when `assertion:` should be used.

Reply:
I've amended with something a little more deterministic - how does it look now?

Reply:
Yeah that's what I meant about hard to distinguish, and impossible would be if you included a normal block without an assertion message. 

Yeah - that's what I suggested about some request information potentially at the log point, that'd give you something more useful in the log.